### PR TITLE
fixed serialization issues with maestro types

### PR DESCRIPTION
**Pull request by SamRoberts**
1. Used Result instead of ValidationNEL in Validator, as scalaz types have problems with kryo
2. Had to disable the accumulation of errors in Validator.all, as Result errors don't combine
3. Used Result instead of \/ in Tag, as scalaz tpyes have problems with kryo
4. Made serializer in LoadEx lazy, so that it doesn't get serialized
5. Moved DateTimeFormatter in Check.isDate outside of the function, and made it lazy, so it
   doesn't have to be recreated for every row, but still doesn't interfere with serialization

NO TESTS YET: I still need to figure out what serialization tests look like. But I wanted to get some feedback on these changes in the meantime, as some of them might be a little controversial.

### CODE REVIEW

File: maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/codec/Tag.scala [hunk 1]
```diff
@@ -30,16 +30,16 @@ object Tag {
...
     Tag(row => {
       if (row.length < fieldsLength)
-        s"Not enough cells in the row. Got ${row.length} expected ${fields.length}.".left
+        Result.fail(s"Not enough cells in the row. Got ${row.length} expected ${fields.length}.")
       else if (row.length > fieldsLength)
-        s"Too many cells in the row. Got ${row.length} expected ${fields.length}.".left
+        Result.fail(s"Too many cells in the row. Got ${row.length} expected ${fields.length}.")
       else
-        (row zip fields).right
+        Result.safe(row zip fields)
```
Comment (stephanh) [inline maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/codec/Tag.scala]: Not `Result.ok`?

Comment (SamRoberts) [inline maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/codec/Tag.scala]: Well, it's a computation inside the brackets. Not one which should fail, in theory, but nonetheless I thought it best to use `Result.safe`.

File: maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/validate/Validator.scala [hunk 2]
```diff
@@ -41,44 +45,44 @@ object Validator {
...
     Validator(a => check.run(field.get(a)).as(a))
 
   /** Creates a [[Validator]] that is always successful. */
-  def pass[A] = Validator[A](a => a.success)
+  def pass[A] = Validator[A](a => Result.ok(a))
 }
 
 object Check {
   def oneOf(categories: String*): Validator[String] =
-    Validator(s => if (categories.contains(s)) s.success else s"""Expected one of [${categories.mkString("|")}] but got ${s}.""".failNel)
```
Comment (stephanh) [inline maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/validate/Validator.scala]: [Codacy](https://www.codacy.com/app/CommBank/maestro/file/458046400/issues/source?bid=2205397&fileBranchId=2235181#l60) File line length exceeds 140 characters

File: maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/validate/Validator.scala [hunk 3]
```diff
@@ -23,15 +25,17 @@ import org.apache.commons.validator.GenericValidator
...
 
 /**
   * Validator to check that the record conforms to expectations.
-  * 
+  *
   * Use this to error out records that don't meet expectations for example
   * checking that records all have a positive balance, etc.
   */
-case class Validator[A](run: A => ValidationNel[String, A])
+case class Validator[A](run: A => Result[A])
```
Comment (stephanh) [inline maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/validate/Validator.scala]: Does the `Validator.all` method need to be changed to capture all the validation errors?

Comment (SamRoberts) [inline maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/validate/Validator.scala]: Yep. This is a bit hard, though. We have a method on Result to add a string to the error, but none to merge arbitrary Results. I was thinking that no-one tended to use multiple Validators, so I wasn't really concerned about losing that functionality.

Comment (SamRoberts) [inline maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/validate/Validator.scala]: I can try to add a combine method to Result if you think it best

Comment (stephanh) [inline maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/validate/Validator.scala]: THat would be good.

**Comment by SamRoberts** [comment]
@stephanh are you happy with this now?

File: maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/validate/Validator.scala [hunk 4]
```diff
@@ -23,62 +25,74 @@ import org.apache.commons.validator.GenericValidator
...
 
   /** Creates a [[Validator]] from a [[Validator]] for a specific field. */
   def of[A, B](field: Field[A, B], check: Validator[B]): Validator[A] =
     Validator(a => check.run(field.get(a)).as(a))
 
   /** Creates a [[Validator]] that is always successful. */
-  def pass[A] = Validator[A](a => a.success)
+  def pass[A] = Validator[A](a => Result.ok(a))
+
+  def combineErrs[A](results: Seq[Result[A]]): Result[Unit] = {
```
Comment (stephanh) [inline maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/validate/Validator.scala]: Scaladoc and spell out `combineErrors`?

**Comment by stephanh** [comment]
:+1:

File: maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/validate/Validator.scala (hunk 4 above)
Comment (SamRoberts) [inline maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/validate/Validator.scala]: Am moving `combineErrs` to a separate object, not part of the maestro API. And spell our Errors as you suggets.

**Comment by stephanh** [comment]
Does this also address #353?

**Comment by SamRoberts** [comment]
Yes, and #142, I think (although that arguably should have been closed when we switched APIs).

**Comment by SamRoberts** [comment]
merged

**Pull request closed without merging**